Api 5393: Include build version in backend

anthony-kopczyk opened the pull request:
# Description

Jira Ticket: https://vajira.max.gov/browse/API-5393

This adds the commit hash of the build into the backend app, accessible at the `/internal/developer-portal/public/version` endpoint. I had to use a public endpoint in order to make it accessible from outside kong.

## Summary of Related Changes

### Code

To get the value of the hash, I use an environment variable (COMMIT_HASH) and return its value. That's really all there is to the code side of things, besides the normal 'set up an endpoint and service'.

### Dockerfile/Build

In order to make the environment variable (COMMIT_HASH) be defined without needing to retrieve a value every time we run the docker container, I added two steps to the Dockerfile.
1. ARG COMMIT_HASH
2. ENV COMMIT_HASH $COMMIT_HASH

The first line `ARG COMMIT_HASH`, defines a build argument for the docker container. The second line sets an environment variable in the container (called `COMMIT_HASH`) that is set to the value of the build argument (also named COMMIT_HASH). These commands make it so the docker image will always have this value on start up, making it consistent and accessible by the backend node app anytime the image is run. Note, that since it's an environment variable, it's value could be overridden at runtime.

In addition to the Dockerfile, I updated the `buildspec-ci.yml` which looks like the only place we run builds right now. It now includes the COMMIT_HASH build argument so that we're passing the hash to the dockerfile.

## Unrelated change that was updated because it was causing frustration

The `docker-compose.yml` was running a build but then running nodemon and ts-node. This was very confusing, so I separated the build section out into it's own docker-compose and left the `docker-compose.yml` without a build. It still runs the same functionally but doesn't do unneeded builds. This should make starting up the backend in the local environment quicker. (No build!). This change also removes the `npm install` call from the docker-compose, so - make sure you run `npm install` on your machine and don't rely on the `docker-compose.yml` to handle it for you.

I added instructions for using the new `docker-compose.build.yml`.

anthony-kopczyk2 committed:
Add commit hash and version via baked env variables

This adds code to bake commit hash and version into our built files.
Since we don't use package.json for versioning, this allows us to inject
these values during the build process within our cicd pipelines.

This generated a file, baking in environment variables kind of like how
React does it. React does a 'search and replace' while we are
overwriting a file and exporting an object that contains the values.

We generate a file 'generated/baked-env.ts' that exports an object
containing all environment variables starting in 'NODE_APP_'. This only
reads variables during the build. So, if we set NODE_APP_EXAMPLE=abc
during the build, then the baked-env.ts will contain NODE_APP_EXAMPLE
and it will be equal to abc.

API-5393

anthony-kopczyk2 committed:
Fix typo - missing end quote

anthony-kopczyk2 committed:
Removed baked-env.ts from source control

This is a generated file, it shouldn't be in source control

anthony-kopczyk2 committed:
Readd baked env to source control

It needs to be in source control for the project to be able to validate
the imports from it are correct.

anthony-kopczyk2 committed:
Add instructions to readme for initial setup

- In the scenario there are no environment variables to bake into the
app, the generate-files script now outputs correctly formatted code.
- Updated README to explain the need to `npm run genfiles` before
running the linter for the first time - explaining the generated files.
- Remove the generated file from source control. I wanted to put an
initial version in source control and then not have it update, but the
gitignore file does not work that way.

API-5393

anthony-kopczyk2 committed:
Fix typoes

In app.test, I had tab characters instead of spaces.

In Dockerfile, I had the wrong environment variables defined for the
generated files.

API-5393

anthony-kopczyk2 committed:
Various fixe and remove baked version from backend

Right now we are focusing on adding the commit hash, so I removed the
baked version.

Side effects of this commit are pretty big.

1. I modified the docker-compose.yml so that it uses the base node12
image rather than the image from the Dockerfile. This keeps it from
needing to run `npm install` and from running `npm run build` which
makes the local start up time much much quicker. Since it's using
ts-node and nodemon still it's quick to reload changes.

2. I added a second docker-compose that retains the functionality of the
original. Used in conjunction with the main docker-compose, this
docker-compose.build.yml ignores the node_modules folder, runs npm
install, and will run npm start - which mimics prod much closer. The
README has been updated with instructions for this Dockerfile

3. The functionality for environment variables may have changed some
from the original concept. Updated the README in the generated folder.
Added comments through files explaining why we're pulling in the
NODE_APP_ arguments in specific places. The basic flow is this:
- When using docker-compose we define an environment variable
NODE_APP_COMMIT_HASH
- Docker-compose takes that environment variable and passes it to
`docker build` as a build argument.
- If not using docker-compose, you just skip and pass it to `docker
build` as a build argument - no need for an environment variable in that
case
- The Dockerfile takes that build argument and adds it as an environment
variable to the docker image.
- When docker runs and the build runs, the commit hash is available via
that environment variable.

API-5393

anthony-kopczyk committed:
Merge branch 'master' into API-5393/include-build-version-in-backend

alexandervalencia reviewed (commented):
Looks good. I'm unsure why the `generate-files.js`and code related to it is necessary? Is there a reason we can't use `process.env` like we are already doing in `configureKongService`?

anthony-kopczyk2 committed:
Pass commit hash as build arg to docker build

The dockerbuild now receives the commit hash making it available to the Dockerfile.
This gives access to the commit hash during `npm run build`

API-5393

anthony-kopczyk2 committed:
Reword generated/README instructions

API-5393

anthony-kopczyk2 committed:
Move 'verson' route to its own handler file

When implementing a route with the commit hash, I originally had it at
the root app route. This isn't available for us when in an environment
because it's not one of the public routes. This commit adds a public
route '/version' and removes the version info from the root path route.

API-5393

anthony-kopczyk2 committed:
Export function getBakedEnv rather than object

This adds an exported function to the baked-env.ts generated file. The
reason is to make it easier to mock with jest in order to test out the
version handler endpoint (routes/Version.ts).

This also adds a test specifically for the baked-env.ts file so that we
can get passable test coverage.

API-5393

anthony-kopczyk2 committed:
Remove generated files

This commit removes the generated files for retrieving the server
version. Instead, since it looks like we use an entire docker image as
our build artifact rather than just the dist/ folder, this commit
relies on the environment variable COMMIT_HASH to present the version.

The Dockerfile uses the ENV command, which permanently sets an
environment variable in the image (this can be overridden if desired).
We were already doing this actually in order to get the environment
variable in the image during the build process - however, it persists to
runtime as well.

The current flow for presenting an environment variable is
1. For local development, add to .env.local or simply set the
environment variable COMMIT_HASH
2. For docker builds, pass COMMIT_HASH as a build argument. The build
argument is passed to the ENV command in the Dockerfile, permanently
setting that environment variable in the image. Again, it can be
overridden if desired, but if it is not, the running docker container
will have the value that was originally passed in as the build argument

API-5393

anthony-kopczyk2 committed:
Remove unneeded instructions from README

With the removal of generated files we no longer need that section of
the README.

API-5393

anthony-kopczyk2 committed:
fix linting problems/typos

API-5393

anthony-kopczyk2 committed:
Cleanup some code

I did some formatting. I also moved the setting of the COMMIT_HASH
environment variable in the Dockerfile to be declared later, after the
npm install.

I updated the comment after looking at setting the PATH.

API-5393

anthony-kopczyk2 committed:
formatting

maddy-jo reviewed (commented):
left a bunch of comments and questions. makes sense in general what you're trying to do with the compose files, but i wanted to clarify some things, and i also had some trouble running it. lmk if i'm missing anything or made a mistake trying to get set up.

anthony-kopczyk2 committed:
Fix typos

Rewords some documentation in the README and made comments in
docker-compose files a little more clear.

Removed an unneeded port on the dynamodb contain in docker-compose.yml

API-5393

anthony-kopczyk2 committed:
Remove VersionService

VersionService was really small with little logic. It wasn't reused
anywhere also, so we took it out

API-5393

anthony-kopczyk2 committed:
Assert when to use docker-compose.build.yml

This adds a little more description to when a developer might want to
use the docker-compose.build.yml and when they might want to not use it.

API-5393

anthony-kopczyk2 committed:
Set COMMIT_HASH environment variable on prod image

The COMMIT_HASH environment variable was only being set in the base
image. This is not carried over to the prod container. THis commit makes
it so we have it in both

API-5393

anthony-kopczyk2 committed:
Fix typo in command name

anthony-kopczyk committed:
Merge branch 'master' into API-5393/include-build-version-in-backend

anthony-kopczyk2 committed:
Install node_modules in local dev container

I removed the node_modules installation in the local development
container when removing an unnecessary call to `npm run build`. After
discussion with other devs, we've decided to readd `npm install` to the
default local dev container.

To do this, I added a new image to the Dockerfile, that contains the npm
install, but does not contain the call to `npm run build`. Updated cicd
pipelines, and updated the docker-compose files.

API-5393

anthony-kopczyk2 committed:
Merge branch 'API-5393/include-build-version-in-backend' of github.com:department-of-veterans-affairs/developer-portal-backend into API-5393/include-build-version-in-backend
